### PR TITLE
Unify ThreadPoolExecutor usage in all Dashboard(Agent)Head.

### PR DESCRIPTION
**Pull request by rynewang**
Previously each Dashboard(Agent)HeadModule can have its own ThreadPoolExecutor. This PR makes a unified TPE in Dashboard(Agent)Head, and uses it everywhere. Also adds asyncio yield in DataOrganizer.organize() to avoid event loop blocking in DataOrganizer.organize for big time.

### CODE REVIEW

**Review by alexeykudinkin — COMMENTED** [review]
@rynewang one more change we'd do is to make frequency of data refresh configurable so that we can mitigate impact by suggesting to users to reduce update frequency

**Comment by rynewang** [comment]
@alexeykudinkin done. Now one can set `RAY_ORGANIZE_DATA_INTERVAL_SECONDS`, defaults to `2`.

File: python/ray/dashboard/datacenter.py [hunk 1]
```diff
@@ -81,7 +90,7 @@ async def organize(cls):
         DataSource.core_worker_stats.reset(core_worker_stats)
```
Comment (jjyao) [inline python/ray/dashboard/datacenter.py:90]: Is DataSource thread safe given now it's accessed in another thread as well

Comment (rynewang) [reply python/ray/dashboard/datacenter.py:90]: not sure...

**Comment by alexeykudinkin** [comment]
PTAL at https://github.com/anyscale/runtime/issues/900#issuecomment-2292570183

**Comment by alexeykudinkin** [comment]
> @alexeykudinkin done. Now one can set RAY_ORGANIZE_DATA_INTERVAL_SECONDS, defaults to 2.

Great! Let's further reduce default to 15s (it'd default for Runway out of the box)

cc @jjyao

File: python/ray/_private/state_api_test_utils.py [hunk 2]
```diff
@@ -324,7 +325,12 @@ def get_state_api_manager(gcs_address: str) -> StateAPIManager:
     state_api_data_source_client = StateDataSourceClient(
         gcs_channel.channel(), gcs_aio_client
     )
-    return StateAPIManager(state_api_data_source_client)
+    return StateAPIManager(
+        state_api_data_source_client,
+        thread_pool_executor=ThreadPoolExecutor(
```
Comment (alexeykudinkin) [inline python/ray/_private/state_api_test_utils.py:330]: Provided that we're planning on offloading CPU-bound tasks that however still hold on to GIL, we should limit # of threads in the TPE (by default TPE provisions at `# of CPUs + 4` threads)

File: python/ray/dashboard/agent.py [hunk 3]
```diff
@@ -47,6 +48,9 @@ def __init__(
         # Public attributes are accessible for all agent modules.
         self.ip = node_ip_address
         self.minimal = minimal
+        self.thread_pool_executor = ThreadPoolExecutor(
```
Comment (alexeykudinkin) [inline python/ray/dashboard/agent.py:51]: Please check my comment above.

We need to
 - Isolate these TPEs to non critical operations (like refreshing stats/data used in UI)
 - Limit concurrency of these TPEs (to 2-4)

File: python/ray/dashboard/consts.py [hunk 4]
```diff
@@ -25,7 +25,7 @@
 RETRY_REDIS_CONNECTION_TIMES = 10
 CONNECT_REDIS_INTERNAL_SECONDS = 2
 PURGE_DATA_INTERVAL_SECONDS = 60 * 10
-ORGANIZE_DATA_INTERVAL_SECONDS = 2
+ORGANIZE_DATA_INTERVAL_SECONDS = env_integer("RAY_ORGANIZE_DATA_INTERVAL_SECONDS", 2)
```
Comment (alexeykudinkin) [inline python/ray/dashboard/consts.py:28]:
```suggestion
ORGANIZE_DATA_INTERVAL_SECONDS = env_integer("RAY_DASHBOARD_STATS_UPDATING_INTERVAL", 2)
```

File: python/ray/dashboard/datacenter.py (hunk 1 above)
Comment (alexeykudinkin) [reply python/ray/dashboard/datacenter.py:90]: +1 

These custom data-structures aren't thread-safe. If we're planning to run this on TPE we need to cover these with locks

File: python/ray/dashboard/head.py [hunk 5]
```diff
@@ -125,6 +125,10 @@ def __init__(
         self._modules_to_load = modules_to_load
         self._modules_loaded = False
 
+        self._thread_pool_executor = ThreadPoolExecutor(
```
Comment (alexeykudinkin) [inline python/ray/dashboard/head.py:128]: Check comment above

**Comment by rynewang** [comment]
todo:
1. revert datacenter.py changes
2. move node_stats change back to async ctx

File: python/ray/_private/state_api_test_utils.py (hunk 2 above)
Comment (rynewang) [reply python/ray/_private/state_api_test_utils.py:330]: changed to 1

Comment (rynewang) [reply python/ray/_private/state_api_test_utils.py:330]: changed in head.py, I mean

**Comment by rynewang** [comment]
@jjyao @alexeykudinkin pls take a look

File: python/ray/dashboard/datacenter.py [hunk 6]
```diff
@@ -67,12 +69,24 @@ async def purge():
...
         node_workers = {}
         core_worker_stats = {}
-        # await inside for loop, so we create a copy of keys().
+        # nodes may change during process, so we create a copy of keys().
         for node_id in list(DataSource.nodes.keys()):
-            workers = await cls.get_node_workers(node_id)
+            # Offloads the blocking operation to a thread pool executor. This also
+            # yields to the event loop.
+            workers = await get_or_create_event_loop().run_in_executor(
+                thread_pool_executor, cls.get_node_workers, node_id
```
Comment (jjyao) [inline python/ray/dashboard/datacenter.py:88]: get_node_workers uses DataSource which is not thread safe, so we cannot just run it in a different thread.


For this PR, let's just call it in the main event loop thread and have yield inside the for loop.

We can optimize `get_node_workers` by stopping using ImmutableDict in a separate PR

File: python/ray/dashboard/head.py [hunk 7]
```diff
@@ -96,6 +97,12 @@ def __init__(
...
+            # We limit number of workers to avoid starving the event loop thread.
```
Comment (jjyao) [inline python/ray/dashboard/head.py:101]: let's explain a bit more why too many worker threads will starve the event loop thread

File: python/ray/dashboard/modules/node/node_head.py [hunk 8]
```diff
@@ -403,31 +404,39 @@ async def _update_node_stats(self):
...
-                    )
-                elif reply.code() == grpc.StatusCode.UNAVAILABLE:
-                    logger.exception(
-                        f"Cannot reach the node, {node_id}. "
-                        "The node may have been terminated."
-                    )
-                else:
+        new_node_stats = {}
+
+        def postprocess():
```
Comment (jjyao) [inline python/ray/dashboard/modules/node/node_head.py:409]: can we make `postprocess` a pure function: accepts (nodes, replies) and returns new_node_stats

**Comment by jjyao** [comment]
After this PR, we should check every call of `dashboard_utils.message_to_dict` and run it in the background thread.

File: python/ray/dashboard/modules/job/job_head.py [hunk 9]
```diff
@@ -317,7 +313,7 @@ async def upload_package(self, req: Request):
         try:
             data = await req.read()
             await get_or_create_event_loop().run_in_executor(
-                self._upload_package_thread_pool_executor,
```
Comment (jjyao) [inline python/ray/dashboard/modules/job/job_head.py:320]: Is it critical enough to deserve its own thread pool?

**Review by jjyao — APPROVED** [review]

File: python/ray/dashboard/modules/job/job_head.py (hunk 9 above)
Comment (rynewang) [reply python/ray/dashboard/modules/job/job_head.py:320]: I don't think so because it's not latency critical either.

**Pull request merged**